src/cargo/util/config/mod.rs

```diff
@@ -262,10 +267,17 @@ impl Config {
 
         let env: HashMap<_, _> = env::vars_os().collect();
 
-        let upper_case_env = env
+        let case_insensitive_env: HashMap<_, _> = env
+            .keys()
+            .filter_map(|k| k.to_str())
+            .map(|k| (k.to_uppercase(), k.to_owned()))
+            .collect();
+
+        let normalized_env = env
             .iter()
-            .filter_map(|(k, _)| k.to_str()) // Only keep valid UTF-8
-            .map(|k| (k.to_uppercase().replace("-", "_"), k.to_owned()))
+            // Only keep entries where both the key and value are valid UTF-8
+            .filter_map(|(k, v)| Some((k.to_str()?, v.to_str()?)))
+            .map(|(k, _)| (k.to_uppercase().replace("-", "_"), k.to_owned()))
```

Comment:
Sorry about that, I meant to add a comment about this but forgot. I realized when I was looking back at #11727 that we might have accidentally subtly changed the behavior of `normalized_env` there.

Before #11727, the `normalized_env` (called `upper_case_env` there) was generated from `self.env`, which held pairs where both the key and value were required to be valid UTF-8. In #11727, `upper_case_env` started using all the keys that were valid UTF-8 (regardless of whether the value was also UTF-8), so in principle more keys might potentially be included.

Since I noticed it here, I put back the check that both the env key and env var are valid UTF-8 before using the key in `normalized_env`. I don't really have a sense of whether this small change is important. If you think that it's probably not, then I'd be glad to use `.keys()` instead, which I'd agree is more readable.